# Please note, all data comes from https://s3.amazonaws.com/tripdata/index.html 

They are very heavy so I won't hold on to the original inputs anymore. However, this script used them, and built the `january_2016_2021_data.csv` file that we will use for the assignment.

In [3]:
import pandas as pd
import numpy as np

In [4]:
bike_2016_df =pd.read_csv('JC-20161-citibike-tripdata.csv')
bike_2016_df.head()

Trip Duration           Start Time            Stop Time  Start Station ID  \
0            362  2016-01-01 00:02:52  2016-01-01 00:08:54              3186   
1            200  2016-01-01 00:18:22  2016-01-01 00:21:42              3186   
2            202  2016-01-01 00:18:25  2016-01-01 00:21:47              3186   
3            248  2016-01-01 00:23:13  2016-01-01 00:27:21              3209   
4            903  2016-01-01 01:03:20  2016-01-01 01:18:24              3195   

  Start Station Name  Start Station Latitude  Start Station Longitude  \
0      Grove St PATH               40.719586               -74.043117   
1      Grove St PATH               40.719586               -74.043117   
2      Grove St PATH               40.719586               -74.043117   
3       Brunswick St               40.724176               -74.050656   
4            Sip Ave               40.730743               -74.063784   

   End Station ID End Station Name  End Station Latitude  \
0            3209     Brunswick St             40.724176   
1            3213   Van Vorst Park             40.718489   
2            3213   Van Vorst Park             40.718489   
3            3203    Hamilton Park             40.727596   
4            3210   Pershing Field             40.742677   

   End Station Longitude  Bike ID   User Type  Birth Year  Gender  
0             -74.050656    24647  Subscriber      1964.0       2  
1             -74.047727    24605  Subscriber      1962.0       1  
2             -74.047727    24689  Subscriber      1962.0       2  
3             -74.044247    24693  Subscriber      1984.0       1  
4             -74.051789    24573    Customer         NaN       0

In [5]:
bike_2017_df = pd.read_csv('JC-201701-citibike-tripdata.csv')
bike_2017_df.head()

Trip Duration           Start Time            Stop Time  Start Station ID  \
0            148  2017-01-01 00:21:32  2017-01-01 00:24:01              3276   
1           1283  2017-01-01 00:24:35  2017-01-01 00:45:58              3183   
2            372  2017-01-01 00:38:19  2017-01-01 00:44:31              3183   
3           1513  2017-01-01 00:38:37  2017-01-01 01:03:50              3194   
4            639  2017-01-01 01:47:52  2017-01-01 01:58:31              3183   

  Start Station Name  Start Station Latitude  Start Station Longitude  \
0   Marin Light Rail               40.714584               -74.042817   
1     Exchange Place               40.716247               -74.033459   
2     Exchange Place               40.716247               -74.033459   
3    McGinley Square               40.725340               -74.067622   
4     Exchange Place               40.716247               -74.033459   

   End Station ID     End Station Name  End Station Latitude  \
0            3185            City Hall             40.717732   
1            3198     Heights Elevator             40.748716   
2            3211           Newark Ave             40.721525   
3            3271  Danforth Light Rail             40.692640   
4            3203        Hamilton Park             40.727596   

   End Station Longitude  Bike ID   User Type  Birth Year  Gender  
0             -74.043845    24575  Subscriber      1983.0       1  
1             -74.040443    24723  Subscriber      1978.0       1  
2             -74.046305    24620  Subscriber      1989.0       1  
3             -74.088012    24668  Subscriber      1961.0       1  
4             -74.044247    26167  Subscriber      1993.0       1

In [6]:
bike_2018_df =pd.read_csv('JC-201801-citibike-tripdata.csv')
bike_2018_df.head()

tripduration                 starttime                  stoptime  \
0           932  2018-01-01 02:06:17.5410  2018-01-01 02:21:50.0270   
1           550  2018-01-01 12:06:18.0390  2018-01-01 12:15:28.4430   
2           510  2018-01-01 12:06:56.9780  2018-01-01 12:15:27.8100   
3           354  2018-01-01 14:53:10.1860  2018-01-01 14:59:05.0960   
4           250  2018-01-01 17:34:30.1920  2018-01-01 17:38:40.9840   

   start station id start station name  start station latitude  \
0              3183     Exchange Place               40.716247   
1              3183     Exchange Place               40.716247   
2              3183     Exchange Place               40.716247   
3              3183     Exchange Place               40.716247   
4              3183     Exchange Place               40.716247   

   start station longitude  end station id end station name  \
0               -74.033459            3199     Newport Pkwy   
1               -74.033459            3199     Newport Pkwy   
2               -74.033459            3199     Newport Pkwy   
3               -74.033459            3267     Morris Canal   
4               -74.033459            3639       Harborside   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.728745             -74.032108   31929  Subscriber   
1             40.728745             -74.032108   31845  Subscriber   
2             40.728745             -74.032108   31708  Subscriber   
3             40.712419             -74.038526   31697  Subscriber   
4             40.719252             -74.034234   31861  Subscriber   

   birth year  gender  
0        1992       1  
1        1969       2  
2        1946       1  
3        1994       1  
4        1991       1

In [7]:
bike_2019_df =pd.read_csv('JC-201901-citibike-tripdata.csv')
bike_2019_df.head()

tripduration                 starttime                  stoptime  \
0           201  2019-01-01 03:09:09.7110  2019-01-01 03:12:30.8790   
1           505  2019-01-01 05:18:00.1060  2019-01-01 05:26:25.9050   
2           756  2019-01-01 10:36:33.3400  2019-01-01 10:49:10.2600   
3          1575  2019-01-01 12:43:38.6430  2019-01-01 13:09:54.5280   
4          1566  2019-01-01 12:43:39.6010  2019-01-01 13:09:46.5100   

   start station id start station name  start station latitude  \
0              3183     Exchange Place               40.716247   
1              3183     Exchange Place               40.716247   
2              3183     Exchange Place               40.716247   
3              3183     Exchange Place               40.716247   
4              3183     Exchange Place               40.716247   

   start station longitude  end station id    end station name  \
0               -74.033459            3214    Essex Light Rail   
1               -74.033459            3638       Washington St   
2               -74.033459            3192  Liberty Light Rail   
3               -74.033459            3638       Washington St   
4               -74.033459            3638       Washington St   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.712774             -74.036486   29612  Subscriber   
1             40.724294             -74.035483   29213  Subscriber   
2             40.711242             -74.055701   26164  Subscriber   
3             40.724294             -74.035483   29672    Customer   
4             40.724294             -74.035483   29522    Customer   

   birth year  gender  
0        1993       1  
1        1972       2  
2        1985       1  
3        1969       0  
4        1969       0

In [8]:
bike_2020_df =pd.read_csv('JC-202001-citibike-tripdata.csv')
bike_2020_df.head()

tripduration                 starttime                  stoptime  \
0           226  2020-01-01 00:04:50.1920  2020-01-01 00:08:37.0370   
1           377  2020-01-01 00:16:01.6700  2020-01-01 00:22:19.0800   
2           288  2020-01-01 00:17:33.8770  2020-01-01 00:22:22.4420   
3           435  2020-01-01 00:32:05.9020  2020-01-01 00:39:21.0660   
4           231  2020-01-01 00:46:19.6780  2020-01-01 00:50:11.3440   

   start station id start station name  start station latitude  \
0              3186      Grove St PATH               40.719586   
1              3186      Grove St PATH               40.719586   
2              3186      Grove St PATH               40.719586   
3              3195            Sip Ave               40.730897   
4              3186      Grove St PATH               40.719586   

   start station longitude  end station id  end station name  \
0               -74.043117            3211        Newark Ave   
1               -74.043117            3269   Brunswick & 6th   
2               -74.043117            3269   Brunswick & 6th   
3               -74.063913            3280       Astor Place   
4               -74.043117            3276  Marin Light Rail   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.721525             -74.046305   29444  Subscriber   
1             40.726012             -74.050389   26305  Subscriber   
2             40.726012             -74.050389   29268    Customer   
3             40.719282             -74.071262   29278    Customer   
4             40.714584             -74.042817   29276  Subscriber   

   birth year  gender  
0        1984       2  
1        1989       2  
2        1989       1  
3        1969       0  
4        1983       2

In [9]:
bike_2021_df =pd.read_csv('JC-202101-citibike-tripdata.csv')
bike_2021_df.head()

tripduration                 starttime                  stoptime  \
0           266  2021-01-01 00:03:35.5100  2021-01-01 00:08:01.7770   
1          1543  2021-01-01 00:23:32.9250  2021-01-01 00:49:16.0830   
2          1461  2021-01-01 00:23:50.7940  2021-01-01 00:48:12.5660   
3           793  2021-01-01 00:31:09.0770  2021-01-01 00:44:22.9430   
4           596  2021-01-01 00:35:52.1900  2021-01-01 00:45:48.7740   

   start station id start station name  start station latitude  \
0              3273       Manila & 1st               40.721651   
1              3681           Grand St               40.715178   
2              3681           Grand St               40.715178   
3              3185          City Hall               40.717732   
4              3639         Harborside               40.719252   

   start station longitude  end station id end station name  \
0               -74.042884            3209     Brunswick St   
1               -74.037683            3213   Van Vorst Park   
2               -74.037683            3213   Van Vorst Park   
3               -74.043845            3199     Newport Pkwy   
4               -74.034234            3209     Brunswick St   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.724176             -74.050656   42494  Subscriber   
1             40.718489             -74.047727   45343    Customer   
2             40.718489             -74.047727   31794    Customer   
3             40.728745             -74.032108   42316    Customer   
4             40.724176             -74.050656   32575    Customer   

   birth year  gender  
0        1988       1  
1        1996       2  
2        1995       1  
3        1969       0  
4        1969       0

BIG SWITCH FROM 2021-2022

In [10]:
bike_2022_df =pd.read_csv('JC-202201-citibike-tripdata.csv')
bike_2022_df.head()

ride_id  rideable_type           started_at             ended_at  \
0  CA5837152804D4B5  electric_bike  2022-01-26 18:50:39  2022-01-26 18:51:53   
1  BA06A5E45B6601D2   classic_bike  2022-01-28 13:14:07  2022-01-28 13:20:23   
2  7B6827D7B9508D93   classic_bike  2022-01-10 19:55:13  2022-01-10 20:00:37   
3  6E5864EA6FCEC90D  electric_bike  2022-01-26 07:54:57  2022-01-26 07:55:22   
4  E24954255BBDE32D  electric_bike  2022-01-13 18:44:46  2022-01-13 18:45:43   

     start_station_name start_station_id      end_station_name end_station_id  \
0  12 St & Sinatra Dr N            HB201  12 St & Sinatra Dr N          HB201   
1      Essex Light Rail            JC038      Essex Light Rail          JC038   
2      Essex Light Rail            JC038      Essex Light Rail          JC038   
3  12 St & Sinatra Dr N            HB201  12 St & Sinatra Dr N          HB201   
4  12 St & Sinatra Dr N            HB201  12 St & Sinatra Dr N          HB201   

   start_lat  start_lng    end_lat    end_lng member_casual  
0  40.750604 -74.024020  40.750604 -74.024020        member  
1  40.712774 -74.036486  40.712774 -74.036486        member  
2  40.712774 -74.036486  40.712774 -74.036486        member  
3  40.750604 -74.024020  40.750604 -74.024020        member  
4  40.750604 -74.024020  40.750604 -74.024020        member

In [11]:
bike_2023_df =pd.read_csv('JC-202301-citibike-tripdata.csv')
bike_2023_df.head()

ride_id  rideable_type           started_at             ended_at  \
0  0905B18B365C9D20   classic_bike  2023-01-28 09:18:10  2023-01-28 09:28:52   
1  B4F0562B05CB5404  electric_bike  2023-01-23 20:10:12  2023-01-23 20:18:27   
2  5ABF032895F5D87E   classic_bike  2023-01-29 15:27:04  2023-01-29 15:32:38   
3  E7E1F9C53976D2F9   classic_bike  2023-01-24 18:35:08  2023-01-24 18:42:13   
4  323165780CA0734B   classic_bike  2023-01-21 20:44:09  2023-01-21 20:48:08   

                         start_station_name start_station_id  \
0  Hoboken Terminal - Hudson St & Hudson Pl            HB101   
1  Hoboken Terminal - Hudson St & Hudson Pl            HB101   
2  Hoboken Terminal - Hudson St & Hudson Pl            HB101   
3  Hoboken Terminal - Hudson St & Hudson Pl            HB101   
4                             Hamilton Park            JC009   

                             end_station_name end_station_id  start_lat  \
0                               Hamilton Park          JC009  40.735938   
1  Southwest Park - Jackson St & Observer Hwy          HB401  40.735938   
2                          Marshall St & 2 St          HB408  40.735944   
3                               Hamilton Park          JC009  40.735986   
4                                Manila & 1st          JC082  40.727596   

   start_lng    end_lat    end_lng member_casual  
0 -74.030305  40.727596 -74.044247        member  
1 -74.030305  40.737551 -74.041664        member  
2 -74.030383  40.740802 -74.042521        member  
3 -74.030364  40.727596 -74.044247        member  
4 -74.044247  40.721651 -74.042884        member

In [12]:
bike_2024_df =pd.read_csv('JC-202401-citibike-tripdata.csv')
bike_2024_df.head()

ride_id  rideable_type           started_at             ended_at  \
0  0744109F13385D1D  electric_bike  2024-01-15 15:18:07  2024-01-15 15:32:44   
1  B1488BFEF9118000   classic_bike  2024-01-13 15:32:50  2024-01-13 15:36:18   
2  95A2FE8E51B4C836   classic_bike  2024-01-19 13:11:00  2024-01-19 13:14:44   
3  95D9AFF6A1652DC1   classic_bike  2024-01-23 07:03:49  2024-01-23 07:07:11   
4  5F7408988A83B1B3   classic_bike  2024-01-01 16:46:10  2024-01-01 16:50:31   

  start_station_name start_station_id end_station_name end_station_id  \
0       Morris Canal            JC072      Oakland Ave          JC022   
1  JC Medical Center            JC110    Grove St PATH          JC115   
2       Morris Canal            JC072      Exchange Pl          JC116   
3       Morris Canal            JC072      Exchange Pl          JC116   
4       Morris Canal            JC072       Harborside          JC104   

   start_lat  start_lng    end_lat    end_lng member_casual  
0  40.712297 -74.038185  40.737604 -74.052478        member  
1  40.715391 -74.049692  40.719410 -74.043090        member  
2  40.712419 -74.038526  40.716366 -74.034344        member  
3  40.712419 -74.038526  40.716366 -74.034344        member  
4  40.712419 -74.038526  40.719252 -74.034234        member

In [13]:
#determine the expected number of rows in the final dataframe for concatenated data from 2016-2021

first_list_bike_dfs = [bike_2016_df, bike_2017_df, bike_2018_df, bike_2019_df, bike_2020_df, bike_2021_df]
num = 0
for df in first_list_bike_dfs:
    num += df.shape[0]

num

90402

In [14]:
#merge two dataframes 
working_stacked_df = pd.concat([bike_2016_df, bike_2017_df], axis=0).reset_index()
working_stacked_df = working_stacked_df.drop(columns='index')
working_stacked_df

Trip Duration           Start Time            Stop Time  \
0                362  2016-01-01 00:02:52  2016-01-01 00:08:54   
1                200  2016-01-01 00:18:22  2016-01-01 00:21:42   
2                202  2016-01-01 00:18:25  2016-01-01 00:21:47   
3                248  2016-01-01 00:23:13  2016-01-01 00:27:21   
4                903  2016-01-01 01:03:20  2016-01-01 01:18:24   
...              ...                  ...                  ...   
20400            329  2017-01-31 23:11:59  2017-01-31 23:17:28   
20401            151  2017-01-31 23:12:04  2017-01-31 23:14:35   
20402            647  2017-01-31 23:19:26  2017-01-31 23:30:13   
20403            295  2017-01-31 23:20:54  2017-01-31 23:25:50   
20404            576  2017-01-31 23:48:37  2017-01-31 23:58:14   

       Start Station ID Start Station Name  Start Station Latitude  \
0                  3186      Grove St PATH               40.719586   
1                  3186      Grove St PATH               40.719586   
2                  3186      Grove St PATH               40.719586   
3                  3209       Brunswick St               40.724176   
4                  3195            Sip Ave               40.730743   
...                 ...                ...                     ...   
20400              3273       Manila & 1st               40.721651   
20401              3185          City Hall               40.717732   
20402              3185          City Hall               40.717732   
20403              3275     Columbus Drive               40.718355   
20404              3201             Dey St               40.737711   

       Start Station Longitude  End Station ID      End Station Name  \
0                   -74.043117            3209          Brunswick St   
1                   -74.043117            3213        Van Vorst Park   
2                   -74.043117            3213        Van Vorst Park   
3                   -74.050656            3203         Hamilton Park   
4                   -74.063784            3210        Pershing Field   
...                        ...             ...                   ...   
20400               -74.042884            3278      Monmouth and 6th   
20401               -74.043845            3276      Marin Light Rail   
20402               -74.043845            3190  Garfield Ave Station   
20403               -74.038914            3214      Essex Light Rail   
20404               -74.066921            3194       McGinley Square   

       End Station Latitude  End Station Longitude  Bike ID   User Type  \
0                 40.724176             -74.050656    24647  Subscriber   
1                 40.718489             -74.047727    24605  Subscriber   
2                 40.718489             -74.047727    24689  Subscriber   
3                 40.727596             -74.044247    24693  Subscriber   
4                 40.742677             -74.051789    24573    Customer   
...                     ...                    ...      ...         ...   
20400             40.725685             -74.048790    26240  Subscriber   
20401             40.714584             -74.042817    24451  Subscriber   
20402             40.710467             -74.070039    26209  Subscriber   
20403             40.712774             -74.036486    26211  Subscriber   
20404             40.725340             -74.067622    26232  Subscriber   

       Birth Year  Gender  
0          1964.0       2  
1          1962.0       1  
2          1962.0       2  
3          1984.0       1  
4             NaN       0  
...           ...     ...  
20400      1980.0       2  
20401      1986.0       1  
20402      1988.0       1  
20403      1988.0       1  
20404      1987.0       1  

[20405 rows x 15 columns]

In [15]:
working_stacked_df.columns = list(bike_2018_df.columns)
working_stacked_df.head()

tripduration            starttime             stoptime  start station id  \
0           362  2016-01-01 00:02:52  2016-01-01 00:08:54              3186   
1           200  2016-01-01 00:18:22  2016-01-01 00:21:42              3186   
2           202  2016-01-01 00:18:25  2016-01-01 00:21:47              3186   
3           248  2016-01-01 00:23:13  2016-01-01 00:27:21              3209   
4           903  2016-01-01 01:03:20  2016-01-01 01:18:24              3195   

  start station name  start station latitude  start station longitude  \
0      Grove St PATH               40.719586               -74.043117   
1      Grove St PATH               40.719586               -74.043117   
2      Grove St PATH               40.719586               -74.043117   
3       Brunswick St               40.724176               -74.050656   
4            Sip Ave               40.730743               -74.063784   

   end station id end station name  end station latitude  \
0            3209     Brunswick St             40.724176   
1            3213   Van Vorst Park             40.718489   
2            3213   Van Vorst Park             40.718489   
3            3203    Hamilton Park             40.727596   
4            3210   Pershing Field             40.742677   

   end station longitude  bikeid    usertype  birth year  gender  
0             -74.050656   24647  Subscriber      1964.0       2  
1             -74.047727   24605  Subscriber      1962.0       1  
2             -74.047727   24689  Subscriber      1962.0       2  
3             -74.044247   24693  Subscriber      1984.0       1  
4             -74.051789   24573    Customer         NaN       0

In [16]:
#merge two dataframes -2016-2017 with 2018
working_stacked_df = pd.concat([working_stacked_df, bike_2018_df], axis=0).reset_index()
working_stacked_df = working_stacked_df.drop(columns='index')
working_stacked_df

tripduration                 starttime                  stoptime  \
0               362       2016-01-01 00:02:52       2016-01-01 00:08:54   
1               200       2016-01-01 00:18:22       2016-01-01 00:21:42   
2               202       2016-01-01 00:18:25       2016-01-01 00:21:47   
3               248       2016-01-01 00:23:13       2016-01-01 00:27:21   
4               903       2016-01-01 01:03:20       2016-01-01 01:18:24   
...             ...                       ...                       ...   
33077           247  2018-01-31 11:03:59.1750  2018-01-31 11:08:07.0570   
33078           806  2018-01-31 15:34:36.8370  2018-01-31 15:48:03.0280   
33079          1499  2018-01-31 15:45:09.8000  2018-01-31 16:10:09.4400   
33080           291  2018-01-31 16:54:11.9480  2018-01-31 16:59:03.5760   
33081           383  2018-01-31 18:34:50.2950  2018-01-31 18:41:13.7530   

       start station id start station name  start station latitude  \
0                  3186      Grove St PATH               40.719586   
1                  3186      Grove St PATH               40.719586   
2                  3186      Grove St PATH               40.719586   
3                  3209       Brunswick St               40.724176   
4                  3195            Sip Ave               40.730743   
...                 ...                ...                     ...   
33077              3640     Journal Square               40.733670   
33078              3640     Journal Square               40.733670   
33079              3640     Journal Square               40.733670   
33080              3640     Journal Square               40.733670   
33081              3640     Journal Square               40.733670   

       start station longitude  end station id end station name  \
0                   -74.043117            3209     Brunswick St   
1                   -74.043117            3213   Van Vorst Park   
2                   -74.043117            3213   Van Vorst Park   
3                   -74.050656            3203    Hamilton Park   
4                   -74.063784            3210   Pershing Field   
...                        ...             ...              ...   
33077               -74.062500            3201           Dey St   
33078               -74.062500            3211       Newark Ave   
33079               -74.062500            3640   Journal Square   
33080               -74.062500            3201           Dey St   
33081               -74.062500            3194  McGinley Square   

       end station latitude  end station longitude  bikeid    usertype  \
0                 40.724176             -74.050656   24647  Subscriber   
1                 40.718489             -74.047727   24605  Subscriber   
2                 40.718489             -74.047727   24689  Subscriber   
3                 40.727596             -74.044247   24693  Subscriber   
4                 40.742677             -74.051789   24573    Customer   
...                     ...                    ...     ...         ...   
33077             40.737711             -74.066921   31828  Subscriber   
33078             40.721525             -74.046305   31842  Subscriber   
33079             40.733670             -74.062500   29622  Subscriber   
33080             40.737711             -74.066921   29622  Subscriber   
33081             40.725340             -74.067622   31690  Subscriber   

       birth year  gender  
0          1964.0       2  
1          1962.0       1  
2          1962.0       2  
3          1984.0       1  
4             NaN       0  
...           ...     ...  
33077      1962.0       1  
33078      1987.0       1  
33079      1981.0       1  
33080      1963.0       1  
33081      1989.0       1  

[33082 rows x 15 columns]

In [17]:
#merge two dataframes -2016-2018 with 2019
working_stacked_df = pd.concat([working_stacked_df, bike_2019_df], axis=0).reset_index()
working_stacked_df = working_stacked_df.drop(columns='index')
working_stacked_df

tripduration                 starttime                  stoptime  \
0               362       2016-01-01 00:02:52       2016-01-01 00:08:54   
1               200       2016-01-01 00:18:22       2016-01-01 00:21:42   
2               202       2016-01-01 00:18:25       2016-01-01 00:21:47   
3               248       2016-01-01 00:23:13       2016-01-01 00:27:21   
4               903       2016-01-01 01:03:20       2016-01-01 01:18:24   
...             ...                       ...                       ...   
52753           853  2019-01-26 06:40:17.3320  2019-01-26 06:54:30.8880   
52754           936  2019-01-27 06:36:20.6400  2019-01-27 06:51:57.1460   
52755           653  2019-01-27 21:14:08.5460  2019-01-27 21:25:01.6170   
52756           960  2019-01-29 06:15:26.5880  2019-01-29 06:31:26.7300   
52757          1109  2019-01-30 06:30:03.6050  2019-01-30 06:48:33.2760   

       start station id start station name  start station latitude  \
0                  3186      Grove St PATH               40.719586   
1                  3186      Grove St PATH               40.719586   
2                  3186      Grove St PATH               40.719586   
3                  3209       Brunswick St               40.724176   
4                  3195            Sip Ave               40.730743   
...                 ...                ...                     ...   
52753              3694     Jackson Square               40.711130   
52754              3694     Jackson Square               40.711130   
52755              3694     Jackson Square               40.711130   
52756              3694     Jackson Square               40.711130   
52757              3694     Jackson Square               40.711130   

       start station longitude  end station id end station name  \
0                   -74.043117            3209     Brunswick St   
1                   -74.043117            3213   Van Vorst Park   
2                   -74.043117            3213   Van Vorst Park   
3                   -74.050656            3203    Hamilton Park   
4                   -74.063784            3210   Pershing Field   
...                        ...             ...              ...   
52753               -74.078900            3183   Exchange Place   
52754               -74.078900            3183   Exchange Place   
52755               -74.078900            3677     Glenwood Ave   
52756               -74.078900            3183   Exchange Place   
52757               -74.078900            3183   Exchange Place   

       end station latitude  end station longitude  bikeid    usertype  \
0                 40.724176             -74.050656   24647  Subscriber   
1                 40.718489             -74.047727   24605  Subscriber   
2                 40.718489             -74.047727   24689  Subscriber   
3                 40.727596             -74.044247   24693  Subscriber   
4                 40.742677             -74.051789   24573    Customer   
...                     ...                    ...     ...         ...   
52753             40.716247             -74.033459   26293  Subscriber   
52754             40.716247             -74.033459   26291  Subscriber   
52755             40.727551             -74.071061   26188    Customer   
52756             40.716247             -74.033459   26196  Subscriber   
52757             40.716247             -74.033459   26242  Subscriber   

       birth year  gender  
0          1964.0       2  
1          1962.0       1  
2          1962.0       2  
3          1984.0       1  
4             NaN       0  
...           ...     ...  
52753      1977.0       1  
52754      1977.0       1  
52755      1998.0       2  
52756      1977.0       1  
52757      1977.0       1  

[52758 rows x 15 columns]

In [18]:
#merge two dataframes -2016-2019 with 2020
working_stacked_df = pd.concat([working_stacked_df, bike_2020_df], axis=0).reset_index()
working_stacked_df = working_stacked_df.drop(columns='index')
working_stacked_df

tripduration                 starttime                  stoptime  \
0               362       2016-01-01 00:02:52       2016-01-01 00:08:54   
1               200       2016-01-01 00:18:22       2016-01-01 00:21:42   
2               202       2016-01-01 00:18:25       2016-01-01 00:21:47   
3               248       2016-01-01 00:23:13       2016-01-01 00:27:21   
4               903       2016-01-01 01:03:20       2016-01-01 01:18:24   
...             ...                       ...                       ...   
78773           544  2020-01-31 23:29:29.3910  2020-01-31 23:38:33.6910   
78774           122  2020-01-31 23:30:59.3670  2020-01-31 23:33:01.6870   
78775           201  2020-01-31 23:42:34.8460  2020-01-31 23:45:55.8780   
78776           300  2020-01-31 23:45:00.6800  2020-01-31 23:50:00.8740   
78777           721  2020-01-31 23:48:35.1700  2020-02-01 00:00:36.4060   

       start station id          start station name  start station latitude  \
0                  3186               Grove St PATH               40.719586   
1                  3186               Grove St PATH               40.719586   
2                  3186               Grove St PATH               40.719586   
3                  3209                Brunswick St               40.724176   
4                  3195                     Sip Ave               40.730743   
...                 ...                         ...                     ...   
78773              3213              Van Vorst Park               40.718489   
78774              3792  Columbus Dr at Exchange Pl               40.716870   
78775              3273                Manila & 1st               40.721651   
78776              3185                   City Hall               40.717732   
78777              3206                     Hilltop               40.731169   

       start station longitude  end station id end station name  \
0                   -74.043117            3209     Brunswick St   
1                   -74.043117            3213   Van Vorst Park   
2                   -74.043117            3213   Van Vorst Park   
3                   -74.050656            3203    Hamilton Park   
4                   -74.063784            3210   Pershing Field   
...                        ...             ...              ...   
78773               -74.047727            3194  McGinley Square   
78774               -74.032810            3639       Harborside   
78775               -74.042884            3209     Brunswick St   
78776               -74.043845            3267     Morris Canal   
78777               -74.057574            3202     Newport PATH   

       end station latitude  end station longitude  bikeid    usertype  \
0                 40.724176             -74.050656   24647  Subscriber   
1                 40.718489             -74.047727   24605  Subscriber   
2                 40.718489             -74.047727   24689  Subscriber   
3                 40.727596             -74.044247   24693  Subscriber   
4                 40.742677             -74.051789   24573    Customer   
...                     ...                    ...     ...         ...   
78773             40.725340             -74.067622   29659  Subscriber   
78774             40.719252             -74.034234   42361  Subscriber   
78775             40.724176             -74.050656   42368  Subscriber   
78776             40.712419             -74.038526   42257  Subscriber   
78777             40.727224             -74.033759   29302    Customer   

       birth year  gender  
0          1964.0       2  
1          1962.0       1  
2          1962.0       2  
3          1984.0       1  
4             NaN       0  
...           ...     ...  
78773      1989.0       1  
78774      1991.0       1  
78775      1988.0       1  
78776      1981.0       2  
78777      1998.0       1  

[78778 rows x 15 columns]

In [19]:
#merge two dataframes -2016-2020 with 2021
working_stacked_df = pd.concat([working_stacked_df, bike_2021_df], axis=0).reset_index()
working_stacked_df = working_stacked_df.drop(columns='index')
working_stacked_df

tripduration                 starttime                  stoptime  \
0               362       2016-01-01 00:02:52       2016-01-01 00:08:54   
1               200       2016-01-01 00:18:22       2016-01-01 00:21:42   
2               202       2016-01-01 00:18:25       2016-01-01 00:21:47   
3               248       2016-01-01 00:23:13       2016-01-01 00:27:21   
4               903       2016-01-01 01:03:20       2016-01-01 01:18:24   
...             ...                       ...                       ...   
90397          2476  2021-01-31 20:16:05.4700  2021-01-31 20:57:22.0110   
90398           147  2021-01-31 21:05:05.2090  2021-01-31 21:07:32.8660   
90399           660  2021-01-31 21:06:23.5880  2021-01-31 21:17:23.8420   
90400           384  2021-01-31 21:16:37.8220  2021-01-31 21:23:02.1240   
90401           356  2021-01-31 23:24:50.8810  2021-01-31 23:30:47.3030   

       start station id start station name  start station latitude  \
0                  3186      Grove St PATH               40.719586   
1                  3186      Grove St PATH               40.719586   
2                  3186      Grove St PATH               40.719586   
3                  3209       Brunswick St               40.724176   
4                  3195            Sip Ave               40.730743   
...                 ...                ...                     ...   
90397              3187          Warren St               40.721124   
90398              3186      Grove St PATH               40.719586   
90399              3211         Newark Ave               40.721525   
90400              3205  JC Medical Center               40.716540   
90401              3639         Harborside               40.719252   

       start station longitude  end station id  end station name  \
0                   -74.043117            3209      Brunswick St   
1                   -74.043117            3213    Van Vorst Park   
2                   -74.043117            3213    Van Vorst Park   
3                   -74.050656            3203     Hamilton Park   
4                   -74.063784            3210    Pershing Field   
...                        ...             ...               ...   
90397               -74.038051            3483     Montgomery St   
90398               -74.043117            3187         Warren St   
90399               -74.046305            3276  Marin Light Rail   
90400               -74.049638            3268    Lafayette Park   
90401               -74.034234            3186     Grove St PATH   

       end station latitude  end station longitude  bikeid    usertype  \
0                 40.724176             -74.050656   24647  Subscriber   
1                 40.718489             -74.047727   24605  Subscriber   
2                 40.718489             -74.047727   24689  Subscriber   
3                 40.727596             -74.044247   24693  Subscriber   
4                 40.742677             -74.051789   24573    Customer   
...                     ...                    ...     ...         ...   
90397             40.719420             -74.050990   44852  Subscriber   
90398             40.721124             -74.038051   46538  Subscriber   
90399             40.714584             -74.042817   44398  Subscriber   
90400             40.713464             -74.062859   41400  Subscriber   
90401             40.719586             -74.043117   43766  Subscriber   

       birth year  gender  
0          1964.0       2  
1          1962.0       1  
2          1962.0       2  
3          1984.0       1  
4             NaN       0  
...           ...     ...  
90397      1958.0       1  
90398      1972.0       1  
90399      1987.0       2  
90400      1989.0       1  
90401      1994.0       1  

[90402 rows x 15 columns]

In [20]:
bike_january_2016_2021_df = working_stacked_df
bike_january_2016_2021_df.to_csv('january_2016_2021_data.csv')

90402 is the expected number of rows, so preliminary check for data loss passses - we should have all the january data across the yeears 2016-2021


And now the 2022 Full year dataframe:

I found this code from here: https://stackoverflow.com/questions/75618004/concatenate-all-csv-files-in-a-folder-using-pandas-and-remove-leading-comma

In [19]:
import glob
from tkinter.filedialog import askdirectory
import os

In [20]:
#create a list of all directory paths:
path_list = ['1_January',
             '2_February',
             '3_March',
             '4_April',
             '5_May',
             '6_June',
             '7_July',
             '8_August',
             '9_September',
             '10_October',
             '11_November',
             '12_December'
]

In [21]:
print(os.getcwd())

c:\Users\rttay\Documents\Education\UTA_Data_Bootcamp\Homeworks_and_Projects\Module_18


In [22]:
os.chdir('2022-citibike-tripdata')

In [23]:
print(os.getcwd())

c:\Users\rttay\Documents\Education\UTA_Data_Bootcamp\Homeworks_and_Projects\Module_18\2022-citibike-tripdata


In [24]:
all_files = []
for path in path_list:
    #navigate to child directory
    os.chdir(path)
    #grab all csv file names
    files = glob.glob("*.csv")
    #construct the file names 
    all_files.append([path + '/' + f for f in files])
    #now we return to parent directory so we can switch child directories
    os.chdir(os.path.abspath('..'))

#flattens the list
all_files = [item for sublist in all_files for item in sublist]
print(all_files)

['1_January/202201-citibike-tripdata_1.csv', '1_January/202201-citibike-tripdata_2.csv', '2_February/202202-citibike-tripdata_1.csv', '2_February/202202-citibike-tripdata_2.csv', '3_March/202203-citibike-tripdata_1.csv', '3_March/202203-citibike-tripdata_2.csv', '4_April/202204-citibike-tripdata_1.csv', '4_April/202204-citibike-tripdata_2.csv', '4_April/202204-citibike-tripdata_3.csv', '5_May/202205-citibike-tripdata_1.csv', '5_May/202205-citibike-tripdata_2.csv', '5_May/202205-citibike-tripdata_3.csv', '6_June/202206-citbike-tripdata_1.csv', '6_June/202206-citbike-tripdata_2.csv', '6_June/202206-citbike-tripdata_3.csv', '6_June/202206-citbike-tripdata_4.csv', '7_July/202207-citbike-tripdata_1.csv', '7_July/202207-citbike-tripdata_2.csv', '7_July/202207-citbike-tripdata_3.csv', '7_July/202207-citbike-tripdata_4.csv', '8_August/202208-citibike-tripdata_1.csv', '8_August/202208-citibike-tripdata_2.csv', '8_August/202208-citibike-tripdata_3.csv', '8_August/202208-citibike-tripdata_4.csv',

In [25]:
#cwd should be 2022-citibike-tripdata

print(os.getcwd())

c:\Users\rttay\Documents\Education\UTA_Data_Bootcamp\Homeworks_and_Projects\Module_18\2022-citibike-tripdata


In [26]:
flist = []

for filename in all_files:
    df = pd.read_csv(filename, index_col=False)
    flist.append(df)

df_out = pd.concat(flist, axis=0, ignore_index=False)
display(df_out.head())


C:\Users\rttay\AppData\Local\Temp\ipykernel_7392\2506200406.py:4: DtypeWarning: Columns (5,7) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(filename, index_col=False)
C:\Users\rttay\AppData\Local\Temp\ipykernel_7392\2506200406.py:4: DtypeWarning: Columns (5,7) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(filename, index_col=False)
C:\Users\rttay\AppData\Local\Temp\ipykernel_7392\2506200406.py:4: DtypeWarning: Columns (5,7) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(filename, index_col=False)
C:\Users\rttay\AppData\Local\Temp\ipykernel_7392\2506200406.py:4: DtypeWarning: Columns (5,7) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(filename, index_col=False)
C:\Users\rttay\AppData\Local\Temp\ipykernel_7392\2506200406.py:4: DtypeWarning: Columns (5,7) have mixed types. Specify dtype option on import o

ride_id  rideable_type           started_at             ended_at  \
0  55262E4365A955A2   classic_bike  2022-01-18 08:23:52  2022-01-18 08:28:18   
1  D272F1B15D841EC0   classic_bike  2022-01-21 09:03:22  2022-01-21 09:05:44   
2  D1FCEF55EB4A807F   classic_bike  2022-01-22 14:28:32  2022-01-22 14:53:18   
3  E9CBDC6A0162C068  electric_bike  2022-01-19 14:49:47  2022-01-19 14:54:02   
4  2177A5B57326CE9B  electric_bike  2022-01-16 14:36:06  2022-01-16 14:44:06   

        start_station_name start_station_id           end_station_name  \
0  Boerum Pl\t& Pacific St          4488.09  Clinton St & Joralemon St   
1          E 12 St & Ave C          5616.08         E 10 St & Avenue A   
2          W 21 St & 6 Ave          6140.05           W 44 St & 11 Ave   
3           38 St & 30 Ave          6850.01       Crescent St & 30 Ave   
4   Pacific St & Nevins St          4362.04    Clinton St & Tillary St   

  end_station_id  start_lat  start_lng    end_lat    end_lng member_casual  
0        4605.04  40.688489 -73.991160  40.692395 -73.993379        member  
1        5659.05  40.727243 -73.976831  40.727408 -73.981420        member  
2        6756.05  40.741740 -73.994156  40.762009 -73.996975        member  
3        6958.06  40.764175 -73.915840  40.768692 -73.924957        member  
4        4748.07  40.685376 -73.983021  40.696233 -73.991421        member

In [27]:
df_out

ride_id  rideable_type           started_at  \
0       55262E4365A955A2   classic_bike  2022-01-18 08:23:52   
1       D272F1B15D841EC0   classic_bike  2022-01-21 09:03:22   
2       D1FCEF55EB4A807F   classic_bike  2022-01-22 14:28:32   
3       E9CBDC6A0162C068  electric_bike  2022-01-19 14:49:47   
4       2177A5B57326CE9B  electric_bike  2022-01-16 14:36:06   
...                  ...            ...                  ...   
649647  CB2EE50F60A58C7B   classic_bike  2022-12-12 08:38:46   
649648  3C0DF3280FEB1350   classic_bike  2022-12-12 13:06:32   
649649  B916E63B630D5020   classic_bike  2022-12-18 01:15:22   
649650  21EAC007BE7D4161   classic_bike  2022-12-08 23:09:54   
649651  538408C4A16F4415  electric_bike  2022-12-21 07:26:14   

                   ended_at                 start_station_name  \
0       2022-01-18 08:28:18            Boerum Pl\t& Pacific St   
1       2022-01-21 09:05:44                    E 12 St & Ave C   
2       2022-01-22 14:53:18                    W 21 St & 6 Ave   
3       2022-01-19 14:54:02                     38 St & 30 Ave   
4       2022-01-16 14:44:06             Pacific St & Nevins St   
...                     ...                                ...   
649647  2022-12-12 08:46:12  E 41 St & Madison Ave (SW corner)   
649648  2022-12-12 13:19:58               Peck Slip & South St   
649649  2022-12-18 01:48:15  E 41 St & Madison Ave (SW corner)   
649650  2022-12-08 23:26:37                   W 37 St & 10 Ave   
649651  2022-12-21 07:31:40                   W 37 St & 10 Ave   

       start_station_id           end_station_name end_station_id  start_lat  \
0               4488.09  Clinton St & Joralemon St        4605.04  40.688489   
1               5616.08         E 10 St & Avenue A        5659.05  40.727243   
2               6140.05           W 44 St & 11 Ave        6756.05  40.741740   
3               6850.01       Crescent St & 30 Ave        6958.06  40.764175   
4               4362.04    Clinton St & Tillary St        4748.07  40.685376   
...                 ...                        ...            ...        ...   
649647          6474.12            6 Ave & W 33 St        6364.07  40.752165   
649648          5096.12       Vesey St & Church St        5216.06  40.707689   
649649          6474.12       Lenox Ave & W 111 St        7602.05  40.752399   
649650          6611.02    W 70 St & Amsterdam Ave        7207.08  40.756604   
649651          6611.02            6 Ave & W 33 St        6364.07  40.756604   

        start_lng    end_lat    end_lng member_casual  
0      -73.991160  40.692395 -73.993379        member  
1      -73.976831  40.727408 -73.981420        member  
2      -73.994156  40.762009 -73.996975        member  
3      -73.915840  40.768692 -73.924957        member  
4      -73.983021  40.696233 -73.991421        member  
...           ...        ...        ...           ...  
649647 -73.979922  40.749013 -73.988484        member  
649648 -74.001254  40.712220 -74.010472        member  
649649 -73.980150  40.798786 -73.952300        member  
649650 -73.997901  40.777480 -73.982886        member  
649651 -73.997901  40.749013 -73.988484        member  

[30689921 rows x 13 columns]

In [28]:
df_out.to_csv('full_2022_trip_data.csv')

As it turns out, that single csv file is upwards of 5 gigs. Without a dedicated sql server, I really can't feasibly work with a data set that large.

I'll stick with the january data from 2016 to 2021. 